File: src/offer34.Path_in_Tree/path_in_tree.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    vector<vector<int> > res;
    vector<int> path;
    void find(TreeNode* root, int sum){
        if (root == nullptr) return;
        path.push_back(root->val);
        sum -= root->val;
        if (sum == 0 && !root->left && !root->right){
            res.push_back(path);
        }
        else{
            if (root->left)
                find(root->left, sum);
            if (root->right)
                find(root->right, sum);
        }
        path.pop_back();
    }
    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {
        
        find(root, expectNumber);
        return res;
    }
};

File: src/offer34.Path_in_Tree/path_in_tree.java
import java.util.ArrayList;
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    ArrayList<ArrayList<Integer>> res = new ArrayList<>();
    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root,int target) {
        helper(new ArrayList<>(), root, target);
        return res;
    }
    private void helper(ArrayList<Integer> list, TreeNode root, int target){
        if(root == null){
            return ;
        }
        list.add(root.val);
        target -= root.val;
        if(target == 0 && root.left == null && root.right == null){
            res.add(new ArrayList<>(list));
        }else{
            helper(list, root.left, target);
            helper(list, root.right, target);
        }
        list.remove(list.size()-1);
    }
}